### DISCUSSION TITLE
NVM Management option?

### DISCUSSION THREAD

**Discussion by GrizzlyAK** [zwave-js/zwave-js-ui#1927] [Q&A] [answered]
Could someone please tell me what the NVM Management (backup/restore) options is under Advanced in the Control Panel? I assume it is to backup/restore the zwave controller, but wanted to ask before doing anything. If so, for an Aeotec ZStick 5+ is this function the same as (i.e., does it replace) using their backup SW?

**Comment by robertsLando** [comment, marked as answer]
I think it's the same. But I let @AlCalzone confirm this

**Reply by AlCalzone** [reply]
AFAIK yes

**Comment by GrizzlyAK** [comment]
OK, thanks guys. BTW, really like the latest changes to the CP. Lots of extra info readily available on-screen. The only thing I would recommend on the current version is the black text on the green background on the Interview status is hard to read. Perhaps white on green, like the check-mark circles? Just an idea. Cheers.

**Comment by GrizzlyAK** [comment]
One the Black on Green issue, that only appears in the Light Theme. The Dark Theme has White on Green.

**Reply by kpine** [reply]
Looks fine to me. Are you sure you're on the latest release?

![image](https://user-images.githubusercontent.com/791794/140662113-b001782a-cd6d-4a0a-8d1d-e182fd0bfac7.png)

**Comment by GrizzlyAK** [comment]
![Untitled-1](https://user-images.githubusercontent.com/4411963/140662308-4a9e88c0-aa9b-47f7-81a5-21db7dd18a2f.gif)

**Reply by kpine** [reply]
5.11.0 is the latest release.

**Reply by kpine** [reply]
Also, configure your S2 keys. 😆

**Comment by GrizzlyAK** [comment]
(1) I think I just had a notification about an update a few days ago, _and updated_. Where do you do a _manual_ update? My Zwave JS to MQTT Add-On is at 0.28.0 and auto-updates and my Integration just gives the ZWave JS Server and Driver versions. I don't even remember how I got the CP (although I think it came with the Add-On).

(2) When I began 2 years ago, security (unless you were using Locks) was more of a performance problem than a benefit, and the advice was to avoid encryption unless you _really_ needed it. With no locks, do you _think_ I really need it? I'm in a fairly sparsely populated area, especially in regards to zwave range, and I'd probably notice someone camped out with a laptop trying to hack my light switches. Am I operating on outdated info?

**Reply by AlCalzone** [reply]
We generally recommend S2, but not S0 unless necessary to operate the device at all. S2 provides a few benefits out of the box, while S0 puts additional stress on the network.

**Reply by kpine** [reply]
The addon is on a different release pace than here, so you you'll have to wait. The theme problem is fixed in 5.11.0 though.

**Comment by GrizzlyAK** [comment]
@kpine ah, ok. No biggie. Thanks.

@AlCalzone OK, I'm going to have to go do some research. My CP does show an S0 Legacy key, although I don't remember ever adding one, nor have I ever added any node to my network with _secure_ (not intentionally). But wait... when I upgraded from OZW, it was all automated (the inclusions), so I don't think I had a choice of No Encryption, which is what I used exclusively in OZW. Since the upgrade, I do believe I have always chosen No Encryption. For those using S0 now, is it as simple as adding an S2 key, and a reinterview, or will it require an exclude/include cycle? For those with No Encryption, I assume an exclude/include?

**Reply by AlCalzone** [reply]
To change the encryption method, you have to exclude and re-include devices, sometimes even factory reset in between.
You can reset a node while zwavejs2mqtt is shut down and use the replace failed node function to reuse the node ID.

**Reply by GrizzlyAK** [reply]
OK, so I'm going to ask for some advice. I have 77 devices, only a handful of which support S2. The docs haven't been updated to reflect the current CP iconography. The grey ? means no encryption? And the red - means S0? Is there any way to know which devices support what from the CP from the icons (or otherwise)? Would this appear in the node's JSON? I want to get a handle on where I am and advice on the best path forward. At this point, I don't know if any my devices are using encryption or not, or how to tell. I assumed they were all unencrypted following the migration.  Any advice would be appreciated. Sorry to be off-topic.

**Reply by kpine** [reply]
Green means secure (tool-tip tells you S0, or which S2 class), red means no security. Grey means "unknown", and I don't know how that happens. Maybe those nodes need to be re-interviewed. Node JSON tells you the same info (grey means the node has no `isSecure` property).

I don't see anything in the node JSON that says whether a node supports S2. The driver logs seem to indicate that based on the message shown at the top of your screenshot. An interview will also show that the node supports the "Security 2" Command Class. The most reliable way to know is if your node comes with a label with either a 5 digit PIN number, or a QR code. Sometimes the label is physically on the device. Sometimes you get a separate card with that info, or it could be in the manual. Next place to check would be the [ZWA DB](https://products.z-wavealliance.org/).

**Reply by GrizzlyAK** [reply]
Nice! Good info. Thanks! 

Yeah, I checked the JSON for that Node 076 referenced by the Driver (top of CP), and it had the following at the very bottom:

  "isSecure": "unknown",
  "supportsSecurity": false,

I still have all the boxes / manuals, so I could figure it out. I guess the question is, _is it worth it for those that do_? And I still don't know which of my devices are using S0 (if any, right?)

**Reply by GrizzlyAK** [reply]
Reinterviewing a couple of the nodes with grey ? didn't change anything.

**Reply by kpine** [reply]
Might be grey because you haven't created the S2 keys.

**Reply by GrizzlyAK** [reply]
Nope. Just did that for all 3 S2 keys and they are still all grey ?. But I did have a Zooz ZEN17 lying around that I added secure (using the default option), which then picked up both S2 Unath and S2 Auth, and which was included successfully. It now has a green check that says S2_Authenticated when hovered.  Yay. My first. ;-)

**Reply by AlCalzone** [reply]
> they are still all grey ?

Did you also re-interview after adding the keys? Guess we need driver logs then.

> I guess the question is, is it worth it for those that do?

You decide. Here are some things S2 supports out of the box:
- Encryption (obviously)
- Message integrity (no more flipped bits resulting in weird sensor measurements)
- Information from the node when it wasn't able to decrypt a command (unlike S0)

Also, with SmartStart, including devices becomes effortless.

The reason we decided to make S2 the default (probably like the Alliance intended) is that you get a bunch of upsides for nearly no downside: A couple of bytes extra on the ether and sometimes re-synchronization of the node with the controller is necessary.

**Reply by GrizzlyAK** [reply]
> Did you also re-interview after adding the keys? 
Yes
> Here are some things S2 supports out of the box:
Wow. That's pretty significant. Beyond the limited NW impact, how much will the en(de)cryption hammer my RPi 4? I idle at about 3% CPU, and the highest I've seen it ever is about 8% when flipping switches on that new S2 device.

For the logs, would it help if I re-interviewed a couple of the (grey ?) devices before dumping the logs and provide the node IDs?

**Reply by AlCalzone** [reply]
> how much will the en(de)cryption hammer my RPi 4

Can't say for sure, but 128 bit AES in C (OpenSSL) is pretty performant.

> For the logs, would it help if I re-interviewed a couple of the (grey ?) devices before dumping the logs and provide the node IDs?

Sure

**Comment by GrizzlyAK** [comment]
Hey, is there any data in either zwave-js or z2m logs that when posted here might reveal security information? I know people sometimes remove their HomeID, for example, but I don't know how that information (or other) could be used by an attacker. I have some logs ready to send/post following a reinterview of a few nodes.

**Reply by AlCalzone** [reply]
IMO the only sensitive information would be user codes for door locks and **maybe** the security keys if you logged an inclusion. The HomeID is a random number that's transmitted unencrypted anyways.

**Comment by GrizzlyAK** [comment]
@robertsLando Just a suggestion, but on the CP, the Advanced box at the top, I _assumed_ that if I had three nodes selected, and did a Re-Interview, that it would only do those three selected nodes, especially since I had it filterd and ONLY those three nodes were showing. My reinterview was taking an especially LONG time, so I removed the filter to find that the command had actually reinterviewed my entire network, all 77 nodes, not just the three selected. You might consider putting a popup warning for that option, or change the behavior.

**Reply by robertsLando** [reply]
> I assumed that if I had three nodes selected,

What you mean with three nodes selected? When you send an action you can actually send it to a single node or broadcast

**Reply by GrizzlyAK** [reply]
Sorry for the delay in responding.

In the CP, I had 3 nodes selected like below. I then chose ReInterview from the top Advanced box (not the one found in each node). I assumed that doing so would only reinterview the 3 nodes I had selected, but it did a network-wide reinterview. It was just not what I expected, given that I could select nodes, and even filter to show only those three.

![image](https://user-images.githubusercontent.com/4411963/141765208-7681cfe3-587b-4139-a7ff-0600579c5145.png)

**Reply by robertsLando** [reply]
@GirzzlyAK Network healing is actually a function that as it says heals the entire network, there is no action that could be sent to just the selected nodes in table. Also there is no function to heal a set of nodes, you should send the heal one by one

**Reply by GrizzlyAK** [reply]
@robertsLando , not Network Healing, I get that, I'm talking about reinterview.

**Reply by AlCalzone** [reply]
The subtitle for "Re-interview Nodes" says "clear **all** info about **all** nodes" though.

**Reply by robertsLando** [reply]
Yeah I mean even with that there is nothing that says the action should be sent only to selected nodes

**Reply by GrizzlyAK** [reply]
OK, like I said, it was just a suggestion. You DID ask me for suggestions on how the CP could be clearer, so there it is. It makes the check boxes redundant if their ONLY function is to filter the items which are displayed, which you can do by filtering the ID column, or any other, already. That's what led me to believe that if there was any _global_ functionality that could apply to a subset of nodes, e.g., those selected, then that's how it would apply. That's why I thought that reinterview would only apply to the rows I selected since, why else did I select and show only them? I did see the subtitle, but that would make sense if no rows were selected.

Am I missing a function that applies to the selected nodes other than just the ones to display (filter)? I get the impression that most HA users don't have more than a screen full of zwave devices anyway. With 77, I make good use of this CP, and the filters (once I figured them out).

No big deal. Now I know. Cheers.

**Reply by AlCalzone** [reply]
If you have specific ideas on what could be improved, opening issues for each idea would be preferred. I don't use it often, but I agree that the control panel is a bit rough around some edges.

**Reply by robertsLando** [reply]
> OK, like I said, it was just a suggestion. You DID ask me for suggestions on how the CP could be clearer, so there it is

Yeah and I'm very open to suggestion @GrizzlyAK , sorry If Iooked rude on my answers I was just tring to explain my point of view :) I agree that lot could be improved on CP, I'm not an UI designer and I prefer to focus on functionalities

**Reply by GrizzlyAK** [reply]
No worries guys. I appreciate all you do to make all of this possible. I recognize how much effort is involved. I get frustrated because I don't have the knowledge currently to do much to help. I did create a new config file today during which in learned a lot. The style guide docs were incredibly helpful. Given the choice I'd vote for functionality too. I mentioned this here instead of as an issue because there are more important things that need your attention.

**Reply by AlCalzone** [reply]
> I mentioned this here instead of as an issue because there are more important things that need your attention.

I'm not sure how familiar you are with Github, but issues can be tracked, organized, labeled and prioritized. So even having low-priority things as issues helps us organize. From my experience, discussions (which I find somewhat chaotic) can easily be forgotten.

**Reply by GrizzlyAK** [reply]
Good to know. Just didn't want to add noise to the stack. Will follow that advice going forward.

I have little experience with GitHub because I've not done any team development in a LONG while. but we used version control systems back in the day when programming C/C++ on Sun Sparcs  (i.e., RCS/SCCS), so I know how it works, just not the details of this particular implementation. But I'm reading up on it and will figure it out.

**Comment by GrizzlyAK** [comment]
@AlCalzone I assume [this guide](https://zwave-js.github.io/node-zwave-js/#/development/intro?id=prerequisites) is tailored to Linux environments? If I want to run VSCode on Win 10, are there extensions that provide the lint / yarn capabilities? Any other recommendations before I install VSCode? I have 3 Linux VMs set up (Ubuntu, Debian, and Mint), but don't use them much. I can still get around Linux, would that be easier/preferred? Any advice would be appreciated.

And have you used Ultraedit Studio with Git integration?

**Reply by robertsLando** [reply]
@GrizzlyAK AFAIK Alcazone uses Windows as development OS, I prefer linux but it's a matter of choice here

**Reply by AlCalzone** [reply]
Like he said, I'm working on Windows 10 myself.

Just install VSCode, [Node.js](https://nodejs.org/en/) and [git for windows](https://gitforwindows.org/). I prefer setting the VSCode terminal to be Git Bash, which behaves just like Bash. Then you can use the same commands as on Linux. Afterwards, you can install yarn with `npm i -g yarn`.
Also make sure to install the recommended extensions in VSCode when it asks you.

**Reply by GrizzlyAK** [reply]
Thanks, @AlCalzone for the details and the links. Do I need to install the Additional Tools included with Node.js, like Chocolatey ? I assume they are _safe_? I have Python 3.9 installed as part of VSCode install, but not sure about the Visual Studio Build Tools (I _do_ have Visual Studio Community 2022 installed). May or may not keep it with VSCode installed.

**Reply by AlCalzone** [reply]
Honestly I've never used Chocolatey, so it's not necessary. The build tools are needed for compiling native packages like serialport. With VS there are tons of different compiler versions so I'd install the included ones to be sure.

**Comment by GrizzlyAK** [comment]
Going back to the security issue, did something change? All of my grey ? are now gone, except for the Aeotec controller. They are now either a red do not enter icon or a green check, where a hover tells you what level of security. I have a question from Zooz about a device of mine, and would like to clarify the following: a red do not enter icon means no security at all (i.e., no S0), and a green check mark can mean S0, S1, and S2, right? Although I never used encryption with OZW, I don't really know how things were added by default (which appears to choose highest available) when I migrated, since I didn't add them one at a time. I assume they are all still unencrypted (not S0), but want to be sure.

**Reply by AlCalzone** [reply]
@robertsLando ?

**Reply by robertsLando** [reply]
Screenshot? BTW about the meaning of icons you should be able to see a tooltop when you hover them with the mouse cursor

**Reply by GrizzlyAK** [reply]
![image](https://user-images.githubusercontent.com/4411963/142373218-27d254c5-43d0-4d99-9dae-837d1ab09a8a.png)

I was hovering over the grey ?. The red says No Security.

**Reply by AlCalzone** [reply]
You can ignore most of the icons for the controller. => https://github.com/zwave-js/zwavejs2mqtt/issues/1886

**Reply by robertsLando** [reply]
I see no error in that, I just have to implement #1886 for controller the rest looks good. Try to hover on the green one you should se the security
